### ISSUE TITLE
Bug/Feature: Intercept Immediately

### ISSUE THREAD

**Issue by avimar** [puppeteer/puppeteer#1378]
**What steps will reproduce the problem?**
I'm using this code:

	async function interceptHandler(target){
		const page = await target.page();
		if(!page) return console.log('target opened: wasn\'t a page');
		console.log('target opened: attaching pre-filter for page');
		await page.setRequestInterception(true);
		page.on('request',blockLoading);
		}

	const browser=await puppeteer.launch(puppeteerOptions);
	browser.on('targetcreated',interceptHandler);
	const page = await browser.newPage();

**What is the expected result?**
Everything blocked by my `blockLoading` function should end up blocked (css/images/fonts).

**What happens instead?**
This works fine for my initial `page` -- probably since there's time to attach the handler.

But when I have a new tab opening, triggering a new target, I see several css/images loading before the intercept handler kicks in to block the rest.

@JoelEinbinder said this is a known issue https://github.com/GoogleChrome/puppeteer/issues/386#issuecomment-343059315 but I couldn't find a ticket.

A possible option from a user's standpoint would be to save a global handler that automatically gets attached to all new targets, so it already exists upon creation if that somehow solves a race issue.

**Comment by ghost** [comment]
Is this issue already taken care? Or is there something known already that can work as a work around to get all the request's data?

**Comment by vc1** [comment]
@nadais  Looks like it can work


https://try-puppeteer.appspot.com/
![image](https://user-images.githubusercontent.com/9018903/38433305-e7618e94-39fc-11e8-85ad-4c6eb2b087db.png)


```javascript
async function interceptHandler(target) {
    const page = await target.page();
    if (!page) return console.log('target opened: wasn\'t a page');
    console.log('target opened: attaching pre-filter for page');
    await page.setRequestInterception(true);
    page.on('request', (request) => {
        console.log(request.resourceType(), request.url())
        request.continue()
    });
}

const browser = await puppeteer.launch(puppeteerOptions);
browser.on('targetcreated', interceptHandler);
const page = await browser.newPage();
await page.goto('https://www.google.com');
```


https://github.com/GoogleChrome/puppeteer/blob/master/examples/block-images.js
https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#event-targetcreated
https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#event-request
https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#class-request

**Comment by ghost** [comment]
That solution also does not work. I tested it in puppeteer with devtools open and the devtools have some requests in the beginning that are not caught by puppeteer

Setup:

1. Click on current page creates a new tab
2. New tab's URL ( let's call it A ) redirects to a URL B and then to a URL C, being URL C the final page URL
3. Devtools detects both URLs A, B and C, puppeteer only detects the URLs that are called within the page loading ( so, inside url C)

**Comment by avimar** [comment]
@vc1 that worked because the `interceptHandler` gets attached to `page` before the `page.goto()` is called.

However, any new `targetcreated` events don't seem to "have time" to run the new handler before requests starts flowing.

**Comment by JoelEinbinder** [comment]
@aslushnikov is it possible for us to have browser-level request interception? It might work well with browser contexts

**Comment by aslushnikov** [comment]
@JoelEinbinder there's been talks about it for quite some time, but I don't think anything like this is feasible unless network service is switched on by default in chromium.

**Comment by berstend** [comment]
I have similar issues without the request interception, I assume for the same reasons:

```es6
'use strict'

const puppeteer = require('puppeteer')
const PUPPETEER_ARGS = ['--no-sandbox', '--disable-setuid-sandbox']

const waitEvent = function (emitter, eventName) {
  return new Promise(resolve => emitter.once(eventName, resolve))
}

puppeteer.launch({ headless: true, args: PUPPETEER_ARGS }).then(async browser => {
  // Modify user agent when page is created
  browser.on('targetcreated', async (target) => {
    if (target.type() !== 'page') { return }
    const page = await target.page()
    let ua = await page.browser().userAgent()
    ua = ua.replace('HeadlessChrome/', 'FunkyChrome/')
    await page.setUserAgent(ua)
  })

  // note: the issue appears with normal context as well
  const context = await browser.createIncognitoBrowserContext()

  // 1st: create a page through pptr
  const page = await context.newPage()
  const ua1 = await page.evaluate(() => window.navigator.userAgent)
  console.assert(!ua1.includes('HeadlessChrome'), '1st')

  // 2nd: navigate that existing target
  await page.goto('about:blank')
  const ua2 = await page.evaluate(() => window.navigator.userAgent)
  console.assert(!ua2.includes('HeadlessChrome'), '2nd')

  // 3rd: implicitly create new target through window.open
  page.evaluate(url => window.open(url), 'http://example.com')
  const popupTarget = await waitEvent(browser, 'targetcreated')
  const popupPage = await popupTarget.page()
  const ua3 = await popupPage.evaluate(() => window.navigator.userAgent)
  console.assert(!ua3.includes('HeadlessChrome'), '3rd')

  await browser.close()
})

// =>
// Assertion failed: 1st
// Assertion failed: 3rd
```

Is there any way to halt newly created pages from loading until the `targetcreated` listeners had sufficient time to register/execute?

I tried even sending `Page.crash` through CDP, but Chrome wouldn't let me 😄 
```
Error: Protocol error (Page.crash): Page has pending navigations, not killing
```

I understand that the 3rd case (the one this ticket is about) is harder to solve, as pptr doesn't know anything about the target until Chrome communicates this. Would still be great to find some sort of workaround here.

My new ticket (#2669) is about the 1st case, where pptr is creating the target but the `targetcreated` event listeners still had no time to do their thing.

**Comment by berstend** [comment]
For reference, I've followed the lifecycle of a popup target:

```
puppeteer:protocol ◀ RECV {"method":"Target.targetCreated","params":{"targetInfo":{"targetId":"05A9FBA8EB66A0B6606600C9970567AB","type":"page","title":"","url":"","attached":false,"openerId":"C564A0CC5E158C4DB940427172992382","browserContextId":"4515C5C5A9995D7F0CA2DE2FE2AE3E74"}}}

puppeteer:protocol ◀ RECV {"method":"Target.targetInfoChanged","params":{"targetInfo":{"targetId":"05A9FBA8EB66A0B6606600C9970567AB","type":"page","title":"httpbin.org/headers","url":"http://httpbin.org/headers","attached":false,"openerId":"C564A0CC5E158C4DB940427172992382","browserContextId":"4515C5C5A9995D7F0CA2DE2FE2AE3E74"}}}

puppeteer:browser browser.emit("targetcreated", Target{"_targetId":"\"05A9FBA8EB66A0B6606600C9970567AB\""})

puppeteer:protocol SEND ► {"id":23,"method":"Target.attachToTarget","params":{"targetId":"05A9FBA8EB66A0B6606600C9970567AB"}} +1ms

puppeteer:protocol ◀ RECV {"method":"Target.targetInfoChanged","params":{"targetInfo":{"targetId":"05A9FBA8EB66A0B6606600C9970567AB","type":"page","title":"httpbin.org/headers","url":"http://httpbin.org/headers","attached":true,"openerId":"C564A0CC5E158C4DB940427172992382","browserContextId":"4515C5C5A9995D7F0CA2DE2FE2AE3E74"}}} +4ms

puppeteer:protocol ◀ RECV {"method":"Target.attachedToTarget","params":{"sessionId":"05A9FBA8EB66A0B6606600C9970567AB:2","targetInfo":{"targetId":"05A9FBA8EB66A0B6606600C9970567AB","type":"page","title":"httpbin.org/headers","url":"http://httpbin.org/headers","attached":true,"openerId":"C564A0CC5E158C4DB940427172992382","browserContextId":"4515C5C5A9995D7F0CA2DE2FE2AE3E74"},"waitingForDebugger":false}} +0ms
```

Could it be a potential problem that `targetcreated` is emitted while the target is not yet attached?

**Comment by berstend** [comment]
Another CDP implementation faced the same issue:
https://github.com/cyrus-and/chrome-remote-interface/issues/319

> Target.setAutoAttach does not work on popups.
> Unfortunately, currently there is no way to attach early enough to not miss first network requests. We are working on it though.

Here is the CDP issue, unfortunately not a priority though:
https://github.com/ChromeDevTools/devtools-protocol/issues/77

**Comment by aslushnikov** [comment]
Closing in favor of #3667.
